chore: support external cert-manager (#1864)
keptn/lifecycle-toolkit: docs/content/en/docs/install/cert-manager.md

Reviewer: should we rename the file to use cert-manager.io? this is the only one we provided instructions for anyhow and we do support it out of the box

Author: we can

Reviewer 2: it's the only widely used one in k8s anyways, AFAIK.
I agree with renaming the page.

Author: renamed

---

> title: Use KLT with cert-manager.io (optional)

Reviewer 3: ```suggestion
title: Use Keptn with cert-manager.io (optional)
```

For a description of the architecture, see
[Keptn Certificate Manager](../concepts/architecture/cert-manager.md).

> KLT also works well with the `cert-manager.io`.
> If you are already `cert-manager.io`,
> you can continue to use it for other components

Reviewer 2: ```suggestion
KLT also works well with `cert-manager.io`.
If you are already using `cert-manager.io`,
you can continue to use it for other components
```

and use the KLT cert-manager just for KLT activities
or you can disable the KLT cert-manager
and configure KLT to use `cert-manager.io`.

If you want KLT to use `cert-manager.io`,
you must configure it *before* you install KLT.
The steps are:

> * Install the `cert-manager.io` if it is not already installed.

Reviewer 2: ```suggestion
* Install `cert-manager.io` if it is not already installed.
```

* Install the cert-manager of your choice
  if it is not already installed.
* Add the `Certificate` and `Issuer` CRs for the cert-manager you are using.

## Add the CRD for your cert-manager

These are the CRs for `cert-manager.io` to be applied to your cluster:

```yaml
apiVersion: cert-manager.io/v1
kind: Certificate
metadata:
  name: klt-certs
  namespace: <your-klt-namespace>
spec:
  dnsNames:

>   - lifecycle-webhook-service.<your-klt-namespace>.svc

Reviewer: don't we need also the validation webhooks? Not sure how this works 🤔

* Install the cert-manager of your choice
  if it is not already installed.

> * Add the `Certificate` and `Issuer` CRs for the cert-manager you are using.

Reviewer 3: ```suggestion
* Add the `Certificate` and `Issuer` CRs for `cert-manager.io`.
```

> * Install KLT without the built-in `klt-cert-manager` via helm

Reviewer 3: ```suggestion
* (optional) Install Keptn without the built-in `klt-cert-manager` via Helm
```

> * (optional) Install Keptn without the built-in `klt-cert-manager` via Helm

Reviewer 4: ```suggestion
* (optional) Disable the built-in  `klt-cert-manager` before installing KLT.
   KLT then looks for the `cert-manager.io` `Certificate` and `Issuer` resources
   and uses those.
```

Author: The disabling of cert-manager happens during KLT installation. I would rather keep this sentence as it is

Reviewer 4: I will defer to you because I don't completely understand this.  But if the cert-manager.io resources are present, what determines whether KLT co-exists with them or uses them instead of the built-in cert-manager?

Author: Basically the existence of Certificate and Issues CRs, this will determine that cert-manager.io will be the one filling up certificates in the secret, which will be used by KLT operators :)

> ## Add the CR(s) for your cert-manager

Reviewer 3: ```suggestion
## Add the CR(s) for cert-manager.io
```

You must modify the KLT manifest for each KLT operator component
to make it aware of the cert-manager you are using.
These instructions implement
[cert-manager.io](https://cert-manager.io/);
the process is similar for other cert-managers.

To configure KLT to use your cert-manager,
change the `Deployment` manifest of each KLT operator component
and **replace** the following `volumes` definition

   ```yaml

>    - emptyDir: {}

Reviewer: so can we leave our empty dir unchanged in the deployments?

Author: yes, we do not need to modify anything here :)

> * Install KLT without `cert-manager` via helm

Reviewer 2: ```suggestion
* Install KLT without the built-in `klt-cert-manager` via helm
```

> ## Add the CRD for your cert-manager

Reviewer 2: ```suggestion
## Add the CRD(s) for your cert-manager
```

>   namespace: <your-klt-namespace>

Reviewer: maybe $NAMESPACE would be more visible? Or we could even suggest a get namespace to save the env var and a command to substitute it and apply it?

Author: I guess this is enough, I mean it's just our good manner that we provide instructions how to run it with external cert-manager, if we already have ours in place

>   namespace: <your-klt-namespace>

Reviewer 3: ```suggestion
  namespace: <your-namespace>
```

>   - lifecycle-webhook-service.<your-klt-namespace>.svc

Reviewer 3: ```suggestion
  - lifecycle-webhook-service.<your-namespace>.svc
```

Reviewer 3: and so forth

>   - lifecycle-webhook-service.<your-klt-namespace>.svc.cluster.local
>   - metrics-webhook-service.<your-klt-namespace>.svc
>   - metrics-webhook-service.<your-klt-namespace>.svc.cluster.local

Author: ```suggestion
  - lifecycle-webhook-service.<your-namespace>.svc.cluster.local
  - metrics-webhook-service.<your-namespace>.svc
  - metrics-webhook-service.<your-namespace>.svc.cluster.local
```

  issuerRef:
    kind: Issuer
    name: klt-selfsigned-issuer
  secretName: klt-certs
---
apiVersion: cert-manager.io/v1
kind: Issuer
metadata:
  name: klt-selfsigned-issuer

>   namespace: <your-klt-namespace>

Author: ```suggestion
  namespace: <your-namespace>
```

spec:
  selfSigned: {}
```

Note the following about these fields:

* The `apiVersion` field refers to the API for the cert-manager.
* The value of the `.spec.secretName` field as well as the `.metadata.name` of the `Certificate` CR

>   must needs to be `klt-certs`.

Reviewer: ```suggestion
  must be `klt-certs`.
```

> * Substitue the namespace placeholders with your namespace, where KLT is installed.

Reviewer 2: ```suggestion
* Substitute the namespace placeholders with your namespace, where KLT is installed.
```

> * Substitute the namespace placeholders with your namespace, where KLT is installed.

Reviewer 3: ```suggestion
* Substitute the namespace placeholders with your namespace, where Keptn is installed.
```
